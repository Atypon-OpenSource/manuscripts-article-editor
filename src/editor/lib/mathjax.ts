// tslint:disable:max-classes-per-file
import { browserAdaptor } from 'mathjax3/mathjax3/adaptors/browserAdaptor'
import { AbstractMathDocument } from 'mathjax3/mathjax3/core/MathDocument'
import { AbstractMathItem } from 'mathjax3/mathjax3/core/MathItem'
import { TeX } from 'mathjax3/mathjax3/input/tex'
import { SVG } from 'mathjax3/mathjax3/output/svg'
import 'mathjax3/mathjax3/util/entities/all'

type Typeset = (math: string | null, display: boolean) => HTMLElement

export interface Mathjax {
  typeset: Typeset
}

class GenericMathDocument<N, T, D> extends AbstractMathDocument<N, T, D> {}
class GenericMathItem<N, T, D> extends AbstractMathItem<N, T, D> {}

const InputJax = new TeX({})
const OutputJax = new SVG()

const doc = new GenericMathDocument(document, browserAdaptor(), { OutputJax })
document.head.appendChild(OutputJax.styleSheet(doc) as Node)

export const typeset = (math: string, display: boolean) => {
  const item = new GenericMathItem(math, InputJax, display)
  item.setMetrics(16, 8, 1000000, 100000, 1)
  item.compile(doc)
  item.typeset(doc)
  return item.typesetRoot
}
