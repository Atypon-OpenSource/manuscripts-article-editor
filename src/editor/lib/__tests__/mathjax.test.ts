import { typeset } from '../mathjax'

describe('mathjax', () => {
  it('generate', () => {
    const tex = '\\sqrt{2}-z_{foo}=\\sum{x}'
    expect(typeset(tex, false)).toMatchSnapshot('mathjax')
    expect(typeset(tex, true)).toMatchSnapshot('mathjax-display')
  })
})
