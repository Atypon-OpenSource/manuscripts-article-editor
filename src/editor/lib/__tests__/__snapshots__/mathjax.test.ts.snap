// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mathjax generate: mathjax 1`] = `
<mjx-container
  class="MathJax"
  jax="SVG"
>
  <svg
    height="3.134ex"
    style="vertical-align: -0.88ex;"
    viewBox="0 -996.5 7273.5 1385.3"
    width="16.456ex"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      fill="currentColor"
      stroke="currentColor"
      stroke-width="0"
      transform="matrix(1 0 0 -1 0 0)"
    >
      <g
        data-mml-node="math"
      >
        <g
          data-mml-node="msqrt"
        >
          <g
            transform="translate(833, 0)"
          >
            <g
              data-mml-node="mn"
            >
              <path
                d="M109 429Q82 429 66 447T50 491Q50 562 103 614T235 666Q326 666 387 610T449 465Q449 422 429 383T381 315T301 241Q265 210 201 149L142 93L218 92Q375 92 385 97Q392 99 409 186V189H449V186Q448 183 436 95T421 3V0H50V19V31Q50 38 56 46T86 81Q115 113 136 137Q145 147 170 174T204 211T233 244T261 278T284 308T305 340T320 369T333 401T340 431T343 464Q343 527 309 573T212 619Q179 619 154 602T119 569T109 550Q109 549 114 549Q132 549 151 535T170 489Q170 464 154 447T109 429Z"
                data-c="32"
              />
            </g>
          </g>
          <g
            data-mml-node="mo"
            transform="translate(0, 55.5)"
          >
            <path
              d="M95 178Q89 178 81 186T72 200T103 230T169 280T207 309Q209 311 212 311H213Q219 311 227 294T281 177Q300 134 312 108L397 -77Q398 -77 501 136T707 565T814 786Q820 800 834 800Q841 800 846 794T853 782V776L620 293L385 -193Q381 -200 366 -200Q357 -200 354 -197Q352 -195 256 15L160 225L144 214Q129 202 113 190T95 178Z"
              data-c="221A"
            />
          </g>
          <rect
            height="60"
            width="500"
            x="833"
            y="795.5"
          />
        </g>
        <g
          data-mml-node="mo"
          transform="translate(1555.2, 0)"
        >
          <path
            d="M84 237T84 250T98 270H679Q694 262 694 250T679 230H98Q84 237 84 250Z"
            data-c="2212"
          />
        </g>
        <g
          data-mml-node="msub"
          transform="translate(2555.4, 0)"
        >
          <g
            data-mml-node="mi"
          >
            <path
              d="M347 338Q337 338 294 349T231 360Q211 360 197 356T174 346T162 335T155 324L153 320Q150 317 138 317Q117 317 117 325Q117 330 120 339Q133 378 163 406T229 440Q241 442 246 442Q271 442 291 425T329 392T367 375Q389 375 411 408T434 441Q435 442 449 442H462Q468 436 468 434Q468 430 463 420T449 399T432 377T418 358L411 349Q368 298 275 214T160 106L148 94L163 93Q185 93 227 82T290 71Q328 71 360 90T402 140Q406 149 409 151T424 153Q443 153 443 143Q443 138 442 134Q425 72 376 31T278 -11Q252 -11 232 6T193 40T155 57Q111 57 76 -3Q70 -11 59 -11H54H41Q35 -5 35 -2Q35 13 93 84Q132 129 225 214T340 322Q352 338 347 338Z"
              data-c="7A"
            />
          </g>
          <g
            data-mml-node="TeXAtom"
            transform="translate(465, -150) scale(0.707)"
          >
            <g
              data-mml-node="mi"
            >
              <path
                d="M118 -162Q120 -162 124 -164T135 -167T147 -168Q160 -168 171 -155T187 -126Q197 -99 221 27T267 267T289 382V385H242Q195 385 192 387Q188 390 188 397L195 425Q197 430 203 430T250 431Q298 431 298 432Q298 434 307 482T319 540Q356 705 465 705Q502 703 526 683T550 630Q550 594 529 578T487 561Q443 561 443 603Q443 622 454 636T478 657L487 662Q471 668 457 668Q445 668 434 658T419 630Q412 601 403 552T387 469T380 433Q380 431 435 431Q480 431 487 430T498 424Q499 420 496 407T491 391Q489 386 482 386T428 385H372L349 263Q301 15 282 -47Q255 -132 212 -173Q175 -205 139 -205Q107 -205 81 -186T55 -132Q55 -95 76 -78T118 -61Q162 -61 162 -103Q162 -122 151 -136T127 -157L118 -162Z"
                data-c="66"
              />
            </g>
            <g
              data-mml-node="mi"
              transform="translate(550, 0)"
            >
              <path
                d="M201 -11Q126 -11 80 38T34 156Q34 221 64 279T146 380Q222 441 301 441Q333 441 341 440Q354 437 367 433T402 417T438 387T464 338T476 268Q476 161 390 75T201 -11ZM121 120Q121 70 147 48T206 26Q250 26 289 58T351 142Q360 163 374 216T388 308Q388 352 370 375Q346 405 306 405Q243 405 195 347Q158 303 140 230T121 120Z"
                data-c="6F"
              />
            </g>
            <g
              data-mml-node="mi"
              transform="translate(1035, 0)"
            >
              <path
                d="M201 -11Q126 -11 80 38T34 156Q34 221 64 279T146 380Q222 441 301 441Q333 441 341 440Q354 437 367 433T402 417T438 387T464 338T476 268Q476 161 390 75T201 -11ZM121 120Q121 70 147 48T206 26Q250 26 289 58T351 142Q360 163 374 216T388 308Q388 352 370 375Q346 405 306 405Q243 405 195 347Q158 303 140 230T121 120Z"
                data-c="6F"
              />
            </g>
          </g>
        </g>
        <g
          data-mml-node="mo"
          transform="translate(4423, 0)"
        >
          <path
            d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"
            data-c="3D"
          />
        </g>
        <g
          data-mml-node="mo"
          transform="translate(5478.8, 0)"
        >
          <path
            d="M61 748Q64 750 489 750H913L954 640Q965 609 976 579T993 533T999 516H979L959 517Q936 579 886 621T777 682Q724 700 655 705T436 710H319Q183 710 183 709Q186 706 348 484T511 259Q517 250 513 244L490 216Q466 188 420 134T330 27L149 -187Q149 -188 362 -188Q388 -188 436 -188T506 -189Q679 -189 778 -162T936 -43Q946 -27 959 6H999L913 -249L489 -250Q65 -250 62 -248Q56 -246 56 -239Q56 -234 118 -161Q186 -81 245 -11L428 206Q428 207 242 462L57 717L56 728Q56 744 61 748Z"
            data-c="2211"
          />
        </g>
        <g
          data-mml-node="TeXAtom"
          transform="translate(6701.5, 0)"
        >
          <g
            data-mml-node="mi"
          >
            <path
              d="M52 289Q59 331 106 386T222 442Q257 442 286 424T329 379Q371 442 430 442Q467 442 494 420T522 361Q522 332 508 314T481 292T458 288Q439 288 427 299T415 328Q415 374 465 391Q454 404 425 404Q412 404 406 402Q368 386 350 336Q290 115 290 78Q290 50 306 38T341 26Q378 26 414 59T463 140Q466 150 469 151T485 153H489Q504 153 504 145Q504 144 502 134Q486 77 440 33T333 -11Q263 -11 227 52Q186 -10 133 -10H127Q78 -10 57 16T35 71Q35 103 54 123T99 143Q142 143 142 101Q142 81 130 66T107 46T94 41L91 40Q91 39 97 36T113 29T132 26Q168 26 194 71Q203 87 217 139T245 247T261 313Q266 340 266 352Q266 380 251 392T217 404Q177 404 142 372T93 290Q91 281 88 280T72 278H58Q52 284 52 289Z"
              data-c="78"
            />
          </g>
        </g>
      </g>
    </g>
  </svg>
</mjx-container>
`;

exports[`mathjax generate: mathjax-display 1`] = `
<mjx-container
  class="MathJax"
  display="true"
  jax="SVG"
>
  <svg
    height="3.634ex"
    style="vertical-align: -1.255ex;"
    viewBox="0 -1051.8 7661.5 1606.3"
    width="17.334ex"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      fill="currentColor"
      stroke="currentColor"
      stroke-width="0"
      transform="matrix(1 0 0 -1 0 0)"
    >
      <g
        data-mml-node="math"
      >
        <g
          data-mml-node="msqrt"
        >
          <g
            transform="translate(833, 0)"
          >
            <g
              data-mml-node="mn"
            >
              <path
                d="M109 429Q82 429 66 447T50 491Q50 562 103 614T235 666Q326 666 387 610T449 465Q449 422 429 383T381 315T301 241Q265 210 201 149L142 93L218 92Q375 92 385 97Q392 99 409 186V189H449V186Q448 183 436 95T421 3V0H50V19V31Q50 38 56 46T86 81Q115 113 136 137Q145 147 170 174T204 211T233 244T261 278T284 308T305 340T320 369T333 401T340 431T343 464Q343 527 309 573T212 619Q179 619 154 602T119 569T109 550Q109 549 114 549Q132 549 151 535T170 489Q170 464 154 447T109 429Z"
                data-c="32"
              />
            </g>
          </g>
          <g
            data-mml-node="mo"
            transform="translate(0, 7.7)"
          >
            <path
              d="M95 178Q89 178 81 186T72 200T103 230T169 280T207 309Q209 311 212 311H213Q219 311 227 294T281 177Q300 134 312 108L397 -77Q398 -77 501 136T707 565T814 786Q820 800 834 800Q841 800 846 794T853 782V776L620 293L385 -193Q381 -200 366 -200Q357 -200 354 -197Q352 -195 256 15L160 225L144 214Q129 202 113 190T95 178Z"
              data-c="221A"
            />
          </g>
          <rect
            height="60"
            width="500"
            x="833"
            y="747.8"
          />
        </g>
        <g
          data-mml-node="mo"
          transform="translate(1555.2, 0)"
        >
          <path
            d="M84 237T84 250T98 270H679Q694 262 694 250T679 230H98Q84 237 84 250Z"
            data-c="2212"
          />
        </g>
        <g
          data-mml-node="msub"
          transform="translate(2555.4, 0)"
        >
          <g
            data-mml-node="mi"
          >
            <path
              d="M347 338Q337 338 294 349T231 360Q211 360 197 356T174 346T162 335T155 324L153 320Q150 317 138 317Q117 317 117 325Q117 330 120 339Q133 378 163 406T229 440Q241 442 246 442Q271 442 291 425T329 392T367 375Q389 375 411 408T434 441Q435 442 449 442H462Q468 436 468 434Q468 430 463 420T449 399T432 377T418 358L411 349Q368 298 275 214T160 106L148 94L163 93Q185 93 227 82T290 71Q328 71 360 90T402 140Q406 149 409 151T424 153Q443 153 443 143Q443 138 442 134Q425 72 376 31T278 -11Q252 -11 232 6T193 40T155 57Q111 57 76 -3Q70 -11 59 -11H54H41Q35 -5 35 -2Q35 13 93 84Q132 129 225 214T340 322Q352 338 347 338Z"
              data-c="7A"
            />
          </g>
          <g
            data-mml-node="TeXAtom"
            transform="translate(465, -150) scale(0.707)"
          >
            <g
              data-mml-node="mi"
            >
              <path
                d="M118 -162Q120 -162 124 -164T135 -167T147 -168Q160 -168 171 -155T187 -126Q197 -99 221 27T267 267T289 382V385H242Q195 385 192 387Q188 390 188 397L195 425Q197 430 203 430T250 431Q298 431 298 432Q298 434 307 482T319 540Q356 705 465 705Q502 703 526 683T550 630Q550 594 529 578T487 561Q443 561 443 603Q443 622 454 636T478 657L487 662Q471 668 457 668Q445 668 434 658T419 630Q412 601 403 552T387 469T380 433Q380 431 435 431Q480 431 487 430T498 424Q499 420 496 407T491 391Q489 386 482 386T428 385H372L349 263Q301 15 282 -47Q255 -132 212 -173Q175 -205 139 -205Q107 -205 81 -186T55 -132Q55 -95 76 -78T118 -61Q162 -61 162 -103Q162 -122 151 -136T127 -157L118 -162Z"
                data-c="66"
              />
            </g>
            <g
              data-mml-node="mi"
              transform="translate(550, 0)"
            >
              <path
                d="M201 -11Q126 -11 80 38T34 156Q34 221 64 279T146 380Q222 441 301 441Q333 441 341 440Q354 437 367 433T402 417T438 387T464 338T476 268Q476 161 390 75T201 -11ZM121 120Q121 70 147 48T206 26Q250 26 289 58T351 142Q360 163 374 216T388 308Q388 352 370 375Q346 405 306 405Q243 405 195 347Q158 303 140 230T121 120Z"
                data-c="6F"
              />
            </g>
            <g
              data-mml-node="mi"
              transform="translate(1035, 0)"
            >
              <path
                d="M201 -11Q126 -11 80 38T34 156Q34 221 64 279T146 380Q222 441 301 441Q333 441 341 440Q354 437 367 433T402 417T438 387T464 338T476 268Q476 161 390 75T201 -11ZM121 120Q121 70 147 48T206 26Q250 26 289 58T351 142Q360 163 374 216T388 308Q388 352 370 375Q346 405 306 405Q243 405 195 347Q158 303 140 230T121 120Z"
                data-c="6F"
              />
            </g>
          </g>
        </g>
        <g
          data-mml-node="mo"
          transform="translate(4423, 0)"
        >
          <path
            d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"
            data-c="3D"
          />
        </g>
        <g
          data-mml-node="mo"
          transform="translate(5478.8, 0)"
        >
          <path
            d="M60 948Q63 950 665 950H1267L1325 815Q1384 677 1388 669H1348L1341 683Q1320 724 1285 761Q1235 809 1174 838T1033 881T882 898T699 902H574H543H251L259 891Q722 258 724 252Q725 250 724 246Q721 243 460 -56L196 -356Q196 -357 407 -357Q459 -357 548 -357T676 -358Q812 -358 896 -353T1063 -332T1204 -283T1307 -196Q1328 -170 1348 -124H1388Q1388 -125 1381 -145T1356 -210T1325 -294L1267 -449L666 -450Q64 -450 61 -448Q55 -446 55 -439Q55 -437 57 -433L590 177Q590 178 557 222T452 366T322 544L56 909L55 924Q55 945 60 948Z"
            data-c="2211"
          />
        </g>
        <g
          data-mml-node="TeXAtom"
          transform="translate(7089.5, 0)"
        >
          <g
            data-mml-node="mi"
          >
            <path
              d="M52 289Q59 331 106 386T222 442Q257 442 286 424T329 379Q371 442 430 442Q467 442 494 420T522 361Q522 332 508 314T481 292T458 288Q439 288 427 299T415 328Q415 374 465 391Q454 404 425 404Q412 404 406 402Q368 386 350 336Q290 115 290 78Q290 50 306 38T341 26Q378 26 414 59T463 140Q466 150 469 151T485 153H489Q504 153 504 145Q504 144 502 134Q486 77 440 33T333 -11Q263 -11 227 52Q186 -10 133 -10H127Q78 -10 57 16T35 71Q35 103 54 123T99 143Q142 143 142 101Q142 81 130 66T107 46T94 41L91 40Q91 39 97 36T113 29T132 26Q168 26 194 71Q203 87 217 139T245 247T261 313Q266 340 266 352Q266 380 251 392T217 404Q177 404 142 372T93 290Q91 281 88 280T72 278H58Q52 284 52 289Z"
              data-c="78"
            />
          </g>
        </g>
      </g>
    </g>
  </svg>
</mjx-container>
`;
